Adds Testing section to Ember.md and standards for writing test scrip…

sernaashui opened the pull request:
…ts in package.json [closes #51]

ASH-Anthony commented on ember.md:
```diff
@@ -183,3 +185,25 @@ CSS is permitted (and encouraged) in apps and addons under certain circumstances
+<a name="testing--test-scripts"></a>
+### 5.1 Test Scripts
+
+> Why? For consistency in testing across all of our Ember apps
```
I think an even more powerful point is: we want meaningful tests and to ensure that those tests pass whenever we build a new version of an app.

ASH-Anthony commented on ember.md:
```diff
@@ -183,3 +185,25 @@ CSS is permitted (and encouraged) in apps and addons under certain circumstances
+
+We use three addons to compile our test scripts.  
+1. `ember-exam` - For allowing more control over which tests run in Ember CLI QUnit
+1. `cross-env` - For allowing single commands for multiple test environments
```
technically not an addon

sernaashui replied:
Dependency?

sernaashui commented:
@ASH-Anthony this is ready for another review.

jduuuubs commented on ember.md:
````diff
@@ -183,3 +185,25 @@ CSS is permitted (and encouraged) in apps and addons under certain circumstances
+``` javascript
+"scripts": {
+    "test": "cross-env COVERAGE=true ember exam",
+    "code-server": "cross-env COVERAGE=true ember exam --server",
````
We never agreed to use a `code-server` NPM script, can you please bring this up in the next meeting if this is something that you would like to advocate the use of?

sernaashui replied:
Will do.

jduuuubs commented on ember.md:
```diff
@@ -183,3 +185,25 @@ CSS is permitted (and encouraged) in apps and addons under certain circumstances
+"scripts": {
+    "test": "cross-env COVERAGE=true ember exam",
+    "code-server": "cross-env COVERAGE=true ember exam --server",
+    "test-server": "cross-env COVERAGE=true ember exam --filter=!acceptance --server"
```
The `test-server` command should not be using exam to filer out acceptance tests. This will prevent these tests from being ran when testing is checked on the build server

sernaashui replied:
That's why there are two, one for `code coverage` and one for `test coverage`. I think only the `code coverage` results should be checked on the build server. We can discuss though.

jduuuubs commented on ember.md:
```diff
@@ -183,3 +185,25 @@ CSS is permitted (and encouraged) in apps and addons under certain circumstances
+
+> Why? To allow us to establish and hold to a standard of code coverage in all of our apps with meaningful test writing and the ability to gate deployments when those standards are not met.
+
+We use three devDependencies to compile our test scripts.  
```
I think it would make more sense to say that we use those dependencies to _run_ our testing scripts

sernaashui commented:
@ASH-Anthony and @mickit I have removed the mention of `ember-exam` until we can discuss its use further in standards and/or team meetings, and I have removed the `code-server` npm test line. I believe this is ready for review again.

ASH-Anthony commented on ember.md:
```diff
@@ -183,3 +185,23 @@ CSS is permitted (and encouraged) in apps and addons under certain circumstances
+> Why? To allow us to establish and hold to a standard of code coverage in all of our apps with meaningful test writing and the ability to gate deployments when those standards are not met.
+
+We use two devDependencies to compile our test scripts.  
+1. `cross-env` - For allowing single commands for multiple test environments
```
the repo says this if for windows usage: https://github.com/kategengler/ember-cli-code-coverage/

> If you want your coverage to work on both Unix and Windows, you can do this:
> 
> npm install cross-env --save-dev

sernaashui replied:
@ASH-Anthony Is there a specific revision that you have in mind?

ASH-Anthony replied:
what do you mean by "multiple test envs?" `cross-env` is needed because we run windows. is that what you're referring to?

sernaashui replied:
I see what you are saying. I can update that verbiage.

sernaashui replied:
Something like:
`cros-env` - For properly setting the NODE_ENV on Windows test environments

ASH-Anthony replied:
i like that

ASH-Bryan replied:
You might want to say `For setting environment variables in a cross-platform manner.` Just because this isn't specific to node.

sernaashui replied:
@ASH-Bryan The previous statement was more general, though. I think this change specifies why we here at ASH use it. Thoughts?

ASH-Bryan replied:
I was just thinking in terms of people understanding that cross-env can be used to set env vars in other contexts - like other scripts they might want to use from package.json.

sernaashui commented:
@ASH-Anthony @ASH-Bryan I believe this is ready to test again.
